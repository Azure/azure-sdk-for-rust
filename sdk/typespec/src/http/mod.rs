// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

mod status_code;

pub use status_code::StatusCode;
