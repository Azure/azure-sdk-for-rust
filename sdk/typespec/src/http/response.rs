// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

//! HTTP responses.

use crate::{
    http::{Headers, StatusCode},
    Bytes,
};
#[cfg(any(feature = "json", feature = "xml"))]
use serde::de::DeserializeOwned;

/// A raw HTTP response with status, headers, and body.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RawResponse {
    status: StatusCode,
    headers: Headers,
    body: Bytes,
}

impl RawResponse {
    /// Create an HTTP response from raw bytes.
    pub fn from_bytes(status: StatusCode, headers: Headers, body: impl Into<Bytes>) -> Self {
        Self {
            status,
            headers,
            body: body.into(),
        }
    }

    /// Get the status code from the response.
    pub fn status(&self) -> StatusCode {
        self.status
    }

    /// Get the headers from the response.
    pub fn headers(&self) -> &Headers {
        &self.headers
    }

    /// Get the response body.
    pub fn body(&self) -> &Bytes {
        &self.body
    }

    /// Get the response body.
    pub fn into_body(self) -> Bytes {
        self.body
    }

    /// Deserialize the JSON stream into type `T`.
    #[cfg(feature = "json")]
    pub fn json<T>(self) -> crate::Result<T>
    where
        T: DeserializeOwned,
    {
        crate::json::from_json(self.body)
    }

    /// Deserialize the XML stream into type `T`.
    #[cfg(feature = "xml")]
    pub fn xml<T>(self) -> crate::Result<T>
    where
        T: DeserializeOwned,
    {
        crate::xml::read_xml(&self.body)
    }
}
