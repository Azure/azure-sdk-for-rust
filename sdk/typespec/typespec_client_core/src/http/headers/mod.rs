// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

//! HTTP headers.

mod common;
mod microsoft;

pub use common::*;
pub use microsoft::*;

use std::{borrow::Cow, fmt::Debug, str::FromStr};
use typespec::error::{Error, ErrorKind, ResultExt};

/// A trait for converting a type into request headers.
pub trait AsHeaders {
    type Iter: Iterator<Item = (HeaderName, HeaderValue)>;
    fn as_headers(&self) -> Self::Iter;
}

impl<T> AsHeaders for T
where
    T: Header,
{
    type Iter = std::vec::IntoIter<(HeaderName, HeaderValue)>;

    /// Iterate over all the header name/value pairs.
    fn as_headers(&self) -> Self::Iter {
        vec![(self.name(), self.value())].into_iter()
    }
}

impl<T> AsHeaders for Option<T>
where
    T: AsHeaders<Iter = std::vec::IntoIter<(HeaderName, HeaderValue)>>,
{
    type Iter = T::Iter;

    /// Iterate over all the header name/value pairs.
    fn as_headers(&self) -> Self::Iter {
        match self {
            Some(h) => h.as_headers(),
            None => vec![].into_iter(),
        }
    }
}

/// View a type as an HTTP header.
///
// Ad interim there are two default functions: `add_to_builder` and `add_to_request`.
//
// While not restricted by the type system, please add HTTP headers only. In particular, do not
// interact with the body of the request.
//
// As soon as the migration to the pipeline architecture will be complete we will phase out
// `add_to_builder`.
pub trait Header {
    fn name(&self) -> HeaderName;
    fn value(&self) -> HeaderValue;
}

/// A collection of headers.
#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub struct Headers(std::collections::HashMap<HeaderName, HeaderValue>);

impl Headers {
    /// Create a new headers collection.
    pub fn new() -> Self {
        Self::default()
    }

    /// Optionally get a header value as a `String`.
    pub fn get_optional_string(&self, key: &HeaderName) -> Option<String> {
        self.get_as(key).ok()
    }

    /// Get a header value as a `str`, or err if it is not found.
    pub fn get_str(&self, key: &HeaderName) -> crate::Result<&str> {
        self.get_with(key, |s| crate::Result::Ok(s.as_str()))
    }

    /// Optionally get a header value as a `str`.
    pub fn get_optional_str(&self, key: &HeaderName) -> Option<&str> {
        self.get_str(key).ok()
    }

    /// Get a header value parsing it as the type, or err if it's not found or it fails to parse.
    pub fn get_as<V, E>(&self, key: &HeaderName) -> crate::Result<V>
    where
        V: FromStr<Err = E>,
        E: std::error::Error + Send + Sync + 'static,
    {
        self.get_with(key, |s| s.as_str().parse())
    }

    /// Optionally get a header value parsing it as the type, or err if it fails to parse.
    pub fn get_optional_as<V, E>(&self, key: &HeaderName) -> crate::Result<Option<V>>
    where
        V: FromStr<Err = E>,
        E: std::error::Error + Send + Sync + 'static,
    {
        self.get_optional_with(key, |s| s.as_str().parse())
    }

    /// Get a header value using the parser, or err if it is not found or fails to parse.
    pub fn get_with<'a, V, F, E>(&'a self, key: &HeaderName, parser: F) -> crate::Result<V>
    where
        F: FnOnce(&'a HeaderValue) -> Result<V, E>,
        E: std::error::Error + Send + Sync + 'static,
    {
        self.get_optional_with(key, parser)?.ok_or_else(|| {
            Error::with_message(ErrorKind::DataConversion, || {
                format!("header not found {}", key.as_str())
            })
        })
    }

    /// Optionally get a header value using the parser, or err if it fails to parse.
    pub fn get_optional_with<'a, V, F, E>(
        &'a self,
        key: &HeaderName,
        parser: F,
    ) -> crate::Result<Option<V>>
    where
        F: FnOnce(&'a HeaderValue) -> Result<V, E>,
        E: std::error::Error + Send + Sync + 'static,
    {
        self.0
            .get(key)
            .map(|v: &HeaderValue| {
                parser(v).with_context(ErrorKind::DataConversion, || {
                    let ty = std::any::type_name::<V>();
                    format!("unable to parse header '{key:?}: {v:?}' into {ty}",)
                })
            })
            .transpose()
    }

    /// Insert a header name/value pair.
    pub fn insert<K, V>(&mut self, key: K, value: V)
    where
        K: Into<HeaderName>,
        V: Into<HeaderValue>,
    {
        self.0.insert(key.into(), value.into());
    }

    /// Add headers to the headers collection.
    pub fn add<H>(&mut self, header: H)
    where
        H: AsHeaders,
    {
        for (key, value) in header.as_headers() {
            self.insert(key, value);
        }
    }

    /// Iterate over all the header name/value pairs.
    pub fn iter(&self) -> impl Iterator<Item = (&HeaderName, &HeaderValue)> {
        self.0.iter()
    }
}

impl IntoIterator for Headers {
    type Item = (HeaderName, HeaderValue);

    type IntoIter = std::collections::hash_map::IntoIter<HeaderName, HeaderValue>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl From<std::collections::HashMap<HeaderName, HeaderValue>> for Headers {
    fn from(c: std::collections::HashMap<HeaderName, HeaderValue>) -> Self {
        Self(c)
    }
}

/// A header name.
#[derive(Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
pub struct HeaderName(Cow<'static, str>);

impl HeaderName {
    /// Create a header name from a static `str`.
    pub const fn from_static(s: &'static str) -> Self {
        ensure_no_uppercase(s);
        Self(Cow::Borrowed(s))
    }

    fn from_cow<C>(c: C) -> Self
    where
        C: Into<Cow<'static, str>>,
    {
        let c = c.into();
        assert!(
            c.chars().all(|c| c.is_lowercase() || !c.is_alphabetic()),
            "header names must be lowercase: {c}"
        );
        Self(c)
    }

    /// Get a header name as a `str`.
    pub fn as_str(&self) -> &str {
        self.0.as_ref()
    }
}

/// Ensures the supplied string does not contain any uppercase ascii characters
const fn ensure_no_uppercase(s: &str) {
    let bytes = s.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        let byte = bytes[i];
        assert!(
            !(byte >= 65u8 && byte <= 90u8),
            "header names must not contain uppercase letters"
        );
        i += 1;
    }
}

impl From<&'static str> for HeaderName {
    fn from(s: &'static str) -> Self {
        Self::from_cow(s)
    }
}

impl From<String> for HeaderName {
    fn from(s: String) -> Self {
        Self::from_cow(s.to_lowercase())
    }
}

/// A header value.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct HeaderValue(Cow<'static, str>);

impl HeaderValue {
    /// Create a header value from a static `str`.
    pub const fn from_static(s: &'static str) -> Self {
        Self(Cow::Borrowed(s))
    }

    /// Create a header value from a [`Cow`].
    pub fn from_cow<C>(c: C) -> Self
    where
        C: Into<Cow<'static, str>>,
    {
        Self(c.into())
    }

    /// Get a header value as a `str`.
    pub fn as_str(&self) -> &str {
        self.0.as_ref()
    }
}

impl From<&'static str> for HeaderValue {
    fn from(s: &'static str) -> Self {
        Self::from_cow(s)
    }
}

impl From<String> for HeaderValue {
    fn from(s: String) -> Self {
        Self::from_cow(s)
    }
}

impl From<&String> for HeaderValue {
    fn from(s: &String) -> Self {
        s.clone().into()
    }
}
