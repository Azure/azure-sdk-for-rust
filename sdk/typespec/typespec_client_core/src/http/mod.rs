// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

//! Types and functions for building HTTP clients.

mod clients;
mod context;
pub mod headers;
mod models;
mod options;
mod pageable;
mod pipeline;
pub mod policies;
pub mod request;
pub mod response;

pub use clients::*;
pub use context::*;
pub use headers::Header;
pub use models::*;
pub use options::*;
pub use pageable::*;
pub use pipeline::*;
pub use request::{Body, Request, RequestContent};
pub use response::{Model, Response};

// Re-export important types.
pub use http_types::{Method, StatusCode};
pub use url::Url;

/// Add a new query pair into the target [`Url`]'s query string.
pub trait AppendToUrlQuery {
    fn append_to_url_query(&self, url: &mut Url);
}

impl<T> AppendToUrlQuery for &T
where
    T: AppendToUrlQuery,
{
    fn append_to_url_query(&self, url: &mut Url) {
        (*self).append_to_url_query(url);
    }
}

impl<T> AppendToUrlQuery for Option<T>
where
    T: AppendToUrlQuery,
{
    fn append_to_url_query(&self, url: &mut Url) {
        if let Some(i) = self {
            i.append_to_url_query(url);
        }
    }
}
