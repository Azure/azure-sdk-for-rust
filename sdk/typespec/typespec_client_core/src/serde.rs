// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

//! serde helpers

/// serde functions for types that are sent and received as strings but aren't surfaced as strings.
pub mod as_string {
    use serde::{Deserialize, Deserializer, Serialize, Serializer};

    /// Deserializes a string into a T. T must implement [`std::str::FromStr`].
    pub fn deserialize<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
    where
        D: Deserializer<'de>,
        T: std::str::FromStr,
        <T as std::str::FromStr>::Err: std::fmt::Display,
    {
        let to_deserialize = <Option<String>>::deserialize(deserializer)?;
        match to_deserialize {
            Some(to_deserialize) => Ok(Some(
                T::from_str(&to_deserialize).map_err(serde::de::Error::custom)?,
            )),
            None => Ok(None),
        }
    }

    /// Serializes T in string format. T must implement [`std::string::ToString`].
    pub fn serialize<S, T>(to_serialize: &Option<T>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        T: std::string::ToString,
    {
        match to_serialize {
            Some(to_serialize) => String::serialize(&to_serialize.to_string(), serializer),
            None => serializer.serialize_none(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::json::{from_json, to_json};
    use serde::{Deserialize, Serialize};
    use serde_json::Number;

    #[derive(Default, Deserialize, Serialize)]
    struct TestType {
        #[serde(
            default,
            skip_serializing_if = "Option::is_none",
            with = "super::as_string"
        )]
        a_bool: Option<bool>,

        #[serde(
            default,
            skip_serializing_if = "Option::is_none",
            with = "super::as_string"
        )]
        json_number: Option<serde_json::Number>,

        #[serde(
            default,
            skip_serializing_if = "Option::is_none",
            with = "super::as_string"
        )]
        some_float: Option<f64>,

        #[serde(
            default,
            skip_serializing_if = "Option::is_none",
            with = "super::as_string"
        )]
        some_int: Option<i32>,
    }

    #[test]
    fn test_deserialize_none() -> crate::Result<()> {
        let json_body = r#"{}"#;
        let test_type: TestType = from_json(json_body)?;
        assert!(test_type.a_bool.is_none());
        assert!(test_type.json_number.is_none());
        assert!(test_type.some_float.is_none());
        assert!(test_type.some_int.is_none());
        Ok(())
    }

    #[test]
    fn test_deserialize_all() -> crate::Result<()> {
        let json_body = r#"{"a_bool":"true","json_number":"123456789","some_float":"3.1415926","some_int":"42"}"#;
        let test_type: TestType = from_json(json_body)?;
        assert_eq!(test_type.a_bool, Some(true));
        assert_eq!(test_type.json_number, Number::from_i128(123456789));
        assert_eq!(test_type.some_float, Some(3.1415926));
        assert_eq!(test_type.some_int, Some(42));
        Ok(())
    }

    #[test]
    fn test_serialize_none() -> crate::Result<()> {
        let test_type = TestType::default();
        let json_body = to_json(&test_type)?;
        assert_eq!(json_body, r#"{}"#);
        Ok(())
    }

    #[test]
    fn test_serialize_all() -> crate::Result<()> {
        let test_type = TestType {
            a_bool: Some(true),
            json_number: Number::from_i128(123456789),
            some_float: Some(3.1415926),
            some_int: Some(42),
        };
        let json_body = to_json(&test_type)?;
        assert_eq!(
            json_body,
            r#"{"a_bool":"true","json_number":"123456789","some_float":"3.1415926","some_int":"42"}"#
        );
        Ok(())
    }
}
