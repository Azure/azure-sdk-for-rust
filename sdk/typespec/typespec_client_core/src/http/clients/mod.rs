// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

//! Built-in HTTP clients.

#[cfg(not(any(feature = "reqwest", feature = "reqwest_rustls")))]
mod noop;
#[cfg(any(feature = "reqwest", feature = "reqwest_rustls"))]
mod reqwest;

#[cfg(not(any(feature = "reqwest", feature = "reqwest_rustls")))]
use self::noop::new_noop_client;
#[cfg(any(feature = "reqwest", feature = "reqwest_rustls"))]
use self::reqwest::new_reqwest_client;

use crate::http::{Request, Response};
use async_trait::async_trait;
use std::sync::Arc;
use typespec::error::Result;

/// Create a new [`HttpClient`].
pub fn new_http_client() -> Arc<dyn HttpClient> {
    #[cfg(any(feature = "reqwest", feature = "reqwest_rustls"))]
    {
        new_reqwest_client()
    }
    #[cfg(not(any(feature = "reqwest", feature = "reqwest_rustls")))]
    {
        new_noop_client()
    }
}

/// An HTTP client which can send requests.
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
pub trait HttpClient: Send + Sync + std::fmt::Debug {
    /// Send out a request using `azure_core`'s types.
    ///
    /// It does not consume the request. Implementors are expected to clone the necessary parts
    /// of the request and pass them to the underlying transport.
    async fn execute_request(&self, request: &Request) -> Result<Response>;
}
