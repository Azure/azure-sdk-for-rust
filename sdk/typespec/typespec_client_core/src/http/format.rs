// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

use crate::http::response::ResponseBody;
use serde::de::DeserializeOwned;

/// A marker trait used to indicate the serialization format used for a response body.
///
/// The [`Response`](crate::http::Response) type uses this trait, in the `F` parameter, to determine how to deserialize the body in to the `T` model when using [`Response::into_body`](crate::http::Response::into_body).
///
/// ## How this trait works
///
/// This trait is a little funky, in order to allow the deserialization behavior of the format to be adjusted based on the type of the response body.
/// This is just a marker trait, it has no methods.
/// Instead, the method to actually perform the deserialization is defined in the [`DeserializeWith`] trait.
/// This trait is parameterized by a type that implements the [`Format`] trait.
pub trait Format: std::fmt::Debug {}

/// A trait used to describe a type that can be deserialized using the specified [`Format`].
///
/// This trait defines the `deserialize_with` method, which takes a [`ResponseBody`] and returns the deserialized value.
/// The `F` type parameter allows for different implementations of the `deserialize_with` method based on the specific [`Format`] marker type used.
///
/// Defining our own trait allows us to implement it on foreign types and better customize deserialization for different scenarios.
#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
pub trait DeserializeWith<F: Format>: Sized {
    async fn deserialize_with(body: ResponseBody) -> typespec::Result<Self>;
}

/// Implements [`DeserializeWith<JsonFormat>`] for an arbitrary type `D`
/// that implements [`serde::de::DeserializeOwned`] by deserializing the response body to the specified type using [`serde_json`].
#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
impl<D: DeserializeOwned> DeserializeWith<JsonFormat> for D {
    async fn deserialize_with(body: ResponseBody) -> typespec::Result<Self> {
        body.json().await
    }
}

/// A [`Format`] that deserializes response bodies using JSON.
/// This is the default format used for deserialization.
///
/// This format supports deserializing response bodies to:
/// * [`ResponseBody`] - The raw response body, without any deserialization.
/// * Any value implementing [`serde::de::DeserializeOwned`] - Deserializes the response body to the specified type using JSON deserialization.
#[derive(Debug, Clone)]
pub struct JsonFormat;

impl Format for JsonFormat {}

/// A [`Format`] that deserializes response bodies using XML.
///
/// This format supports deserializing response bodies to:
/// * [`ResponseBody`] - The raw response body, without any deserialization.
/// * Any value implementing [`serde::de::DeserializeOwned`] - Deserializes the response body to the specified type using XML deserialization.
#[cfg(feature = "xml")]
#[derive(Debug, Clone)]
pub struct XmlFormat;

#[cfg(feature = "xml")]
impl Format for XmlFormat {}

#[cfg(feature = "xml")]
#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
impl<D: DeserializeOwned> DeserializeWith<XmlFormat> for D {
    async fn deserialize_with(body: ResponseBody) -> typespec::Result<Self> {
        body.xml().await
    }
}

/// A [`Format`] indicating that the response has no structured format.
/// This includes responses that return raw data and that don't return a response body.
///
/// This format supports deserializing response bodies to:
/// * [`ResponseBody`] - The raw response body, without any deserialization.
#[derive(Debug, Clone)]
pub struct NoFormat;

impl Format for NoFormat {}
