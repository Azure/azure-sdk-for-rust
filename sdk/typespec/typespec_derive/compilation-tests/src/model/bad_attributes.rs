// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

use serde::Deserialize;
use typespec_client_core::http::Model;

#[derive(Model, Deserialize)]
#[typespec(foobar)]
pub struct NotAValidAttribute {}

#[derive(Model, Deserialize)]
#[typespec(crate = 42)]
pub struct NumericLiteralCrate {}

#[derive(Model, Deserialize)]
#[typespec(crate = "a" + "b")]
pub struct BinExprCrate {}

#[derive(Model, Deserialize)]
#[typespec(crate = @)]
pub struct UnexpectedTokenCrate {}

#[derive(Model, Deserialize)]
#[typespec(crate = "a b c")]
pub struct InvalidPathOnCrate {}

#[derive(Model, Deserialize)]
#[typespec(format = 42)]
pub struct NotAStringLiteralOnFormat {}

#[derive(Model, Deserialize)]
#[typespec(format("json"))]
pub struct IncorrectAttributeFormat {}

#[derive(Model, Deserialize)]
#[typespec = "whoop"]
pub struct NotAMetaListAttribute {}
