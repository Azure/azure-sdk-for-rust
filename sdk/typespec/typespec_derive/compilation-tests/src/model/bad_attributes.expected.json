[
  {
    "package_id": "<redacted>",
    "target": {
      "name": "compilation-tests",
      "kind": [
        "lib"
      ],
      "crate_types": [
        "lib"
      ],
      "required-features": [],
      "src_path": "sdk/typespec/typespec_derive/compilation-tests/src/lib.rs",
      "edition": "2021",
      "doctest": true,
      "test": true,
      "doc": true
    },
    "message": {
      "message": "unknown typespec attribute '#[typespec(blork)'",
      "code": null,
      "level": "error",
      "spans": [
        {
          "file_name": "sdk/typespec/typespec_derive/compilation-tests/src/model/bad_attributes.rs",
          "byte_start": 105,
          "byte_end": 110,
          "line_start": 5,
          "line_end": 5,
          "column_start": 12,
          "column_end": 17,
          "is_primary": true,
          "text": [
            {
              "text": "#[typespec(blork)]",
              "highlight_start": 12,
              "highlight_end": 17
            }
          ],
          "label": null,
          "suggested_replacement": null,
          "suggestion_applicability": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": "error: unknown typespec attribute '#[typespec(blork)'\n --> src/model/bad_attributes.rs:5:12\n  |\n5 | #[typespec(blork)]\n  |            ^^^^^\n\n"
    }
  },
  {
    "package_id": "<redacted>",
    "target": {
      "name": "compilation-tests",
      "kind": [
        "lib"
      ],
      "crate_types": [
        "lib"
      ],
      "required-features": [],
      "src_path": "sdk/typespec/typespec_derive/compilation-tests/src/lib.rs",
      "edition": "2021",
      "doctest": true,
      "test": true,
      "doc": true
    },
    "message": {
      "message": "expected string literal",
      "code": null,
      "level": "error",
      "spans": [
        {
          "file_name": "sdk/typespec/typespec_derive/compilation-tests/src/model/bad_attributes.rs",
          "byte_start": 198,
          "byte_end": 199,
          "line_start": 9,
          "line_end": 9,
          "column_start": 22,
          "column_end": 23,
          "is_primary": true,
          "text": [
            {
              "text": "#[typespec(crate = 42)]",
              "highlight_start": 22,
              "highlight_end": 23
            }
          ],
          "label": null,
          "suggested_replacement": null,
          "suggestion_applicability": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": "error: expected string literal\n --> src/model/bad_attributes.rs:9:22\n  |\n9 | #[typespec(crate = 42)]\n  |                      ^\n\n"
    }
  },
  {
    "package_id": "<redacted>",
    "target": {
      "name": "compilation-tests",
      "kind": [
        "lib"
      ],
      "crate_types": [
        "lib"
      ],
      "required-features": [],
      "src_path": "sdk/typespec/typespec_derive/compilation-tests/src/lib.rs",
      "edition": "2021",
      "doctest": true,
      "test": true,
      "doc": true
    },
    "message": {
      "message": "expected string literal",
      "code": null,
      "level": "error",
      "spans": [
        {
          "file_name": "sdk/typespec/typespec_derive/compilation-tests/src/model/bad_attributes.rs",
          "byte_start": 294,
          "byte_end": 295,
          "line_start": 13,
          "line_end": 13,
          "column_start": 29,
          "column_end": 30,
          "is_primary": true,
          "text": [
            {
              "text": "#[typespec(crate = \"a\" + \"b\")]",
              "highlight_start": 29,
              "highlight_end": 30
            }
          ],
          "label": null,
          "suggested_replacement": null,
          "suggestion_applicability": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": "error: expected string literal\n  --> src/model/bad_attributes.rs:13:29\n   |\n13 | #[typespec(crate = \"a\" + \"b\")]\n   |                             ^\n\n"
    }
  },
  {
    "package_id": "<redacted>",
    "target": {
      "name": "compilation-tests",
      "kind": [
        "lib"
      ],
      "crate_types": [
        "lib"
      ],
      "required-features": [],
      "src_path": "sdk/typespec/typespec_derive/compilation-tests/src/lib.rs",
      "edition": "2021",
      "doctest": true,
      "test": true,
      "doc": true
    },
    "message": {
      "message": "expected string literal",
      "code": null,
      "level": "error",
      "spans": [
        {
          "file_name": "sdk/typespec/typespec_derive/compilation-tests/src/model/bad_attributes.rs",
          "byte_start": 374,
          "byte_end": 375,
          "line_start": 17,
          "line_end": 17,
          "column_start": 20,
          "column_end": 21,
          "is_primary": true,
          "text": [
            {
              "text": "#[typespec(crate = @)]",
              "highlight_start": 20,
              "highlight_end": 21
            }
          ],
          "label": null,
          "suggested_replacement": null,
          "suggestion_applicability": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": "error: expected string literal\n  --> src/model/bad_attributes.rs:17:20\n   |\n17 | #[typespec(crate = @)]\n   |                    ^\n\n"
    }
  },
  {
    "package_id": "<redacted>",
    "target": {
      "name": "compilation-tests",
      "kind": [
        "lib"
      ],
      "crate_types": [
        "lib"
      ],
      "required-features": [],
      "src_path": "sdk/typespec/typespec_derive/compilation-tests/src/lib.rs",
      "edition": "2021",
      "doctest": true,
      "test": true,
      "doc": true
    },
    "message": {
      "message": "invalid module path: a b c",
      "code": null,
      "level": "error",
      "spans": [
        {
          "file_name": "sdk/typespec/typespec_derive/compilation-tests/src/model/bad_attributes.rs",
          "byte_start": 463,
          "byte_end": 470,
          "line_start": 21,
          "line_end": 21,
          "column_start": 20,
          "column_end": 27,
          "is_primary": true,
          "text": [
            {
              "text": "#[typespec(crate = \"a b c\")]",
              "highlight_start": 20,
              "highlight_end": 27
            }
          ],
          "label": null,
          "suggested_replacement": null,
          "suggestion_applicability": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": "error: invalid module path: a b c\n  --> src/model/bad_attributes.rs:21:20\n   |\n21 | #[typespec(crate = \"a b c\")]\n   |                    ^^^^^^^\n\n"
    }
  },
  {
    "package_id": "<redacted>",
    "target": {
      "name": "compilation-tests",
      "kind": [
        "lib"
      ],
      "crate_types": [
        "lib"
      ],
      "required-features": [],
      "src_path": "sdk/typespec/typespec_derive/compilation-tests/src/lib.rs",
      "edition": "2021",
      "doctest": true,
      "test": true,
      "doc": true
    },
    "message": {
      "message": "expected string literal",
      "code": null,
      "level": "error",
      "spans": [
        {
          "file_name": "sdk/typespec/typespec_derive/compilation-tests/src/model/bad_attributes.rs",
          "byte_start": 559,
          "byte_end": 560,
          "line_start": 25,
          "line_end": 25,
          "column_start": 23,
          "column_end": 24,
          "is_primary": true,
          "text": [
            {
              "text": "#[typespec(format = 42)]",
              "highlight_start": 23,
              "highlight_end": 24
            }
          ],
          "label": null,
          "suggested_replacement": null,
          "suggestion_applicability": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": "error: expected string literal\n  --> src/model/bad_attributes.rs:25:23\n   |\n25 | #[typespec(format = 42)]\n   |                       ^\n\n"
    }
  },
  {
    "package_id": "<redacted>",
    "target": {
      "name": "compilation-tests",
      "kind": [
        "lib"
      ],
      "crate_types": [
        "lib"
      ],
      "required-features": [],
      "src_path": "sdk/typespec/typespec_derive/compilation-tests/src/lib.rs",
      "edition": "2021",
      "doctest": true,
      "test": true,
      "doc": true
    },
    "message": {
      "message": "invalid typespec attribute, expected attribute in form #[typespec(key = value)]",
      "code": null,
      "level": "error",
      "spans": [
        {
          "file_name": "sdk/typespec/typespec_derive/compilation-tests/src/model/bad_attributes.rs",
          "byte_start": 633,
          "byte_end": 634,
          "line_start": 29,
          "line_end": 29,
          "column_start": 1,
          "column_end": 2,
          "is_primary": true,
          "text": [
            {
              "text": "#[typespec = \"whoop\"]",
              "highlight_start": 1,
              "highlight_end": 2
            }
          ],
          "label": null,
          "suggested_replacement": null,
          "suggestion_applicability": null,
          "expansion": null
        }
      ],
      "children": [],
      "rendered": "error: invalid typespec attribute, expected attribute in form #[typespec(key = value)]\n  --> src/model/bad_attributes.rs:29:1\n   |\n29 | #[typespec = \"whoop\"]\n   | ^\n\n"
    }
  }
]